File: Week_03/22/generate_parentheses.py
from typing import List


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        solution = [""] * 2 * n

        def dfs(left: int, right: int):
            if right == n:
                res.append("".join(solution))
                return
            if left < n:
                solution[left + right] = "("
                dfs(left + 1, right)
            if right < left:
                solution[left + right] = ")"
                dfs(left, right + 1)

        dfs(0, 0)
        return res

File: Week_03/22/generate_parentheses.go
package main

func generateParenthesis(n int) []string {
	var res []string
	solution := make([]byte, n*2)
	var dfs func(left, right int)
	dfs = func(left, right int) {
		if right == n {
			res = append(res, string(solution))
			return
		}
		if left < n {
			solution[left+right] = '('
			dfs(left+1, right)
		}
		if right < left {
			solution[left+right] = ')'
			dfs(left, right+1)
		}
	}
	dfs(0, 0)
	return res
}
